### PR TITLE
fix(s13a.13): self-ref look-back no-prior short-circuit — Phase 6 #3f

### LINKED ISSUE #68: S13a.13: `a = ${?a}foo` resolves to "foofoo" instead of "foo"
## Spec reference

HOCON.md L841 (§Self-Referential substitutions):

> `a = ${?a}foo` should resolve to `"foo"` when `a` has no prior value, because the optional self-reference `${?a}` looks back and finds nothing, so it contributes an empty string to the concatenation.

## Current behaviour

```
a = ${?a}foo
```

Resolves to `"foofoo"` — i.e. the resolver resolves `${?a}` to `"foo"` (the value being assembled) rather than treating it as absent.

## Expected behaviour

`a` should equal `"foo"` — the optional look-back `${?a}` finds no prior value and contributes nothing.

## Root cause (hypothesis)

The self-reference detection in `resolver.go` (around `resolveSubst`) detects that `a` is self-referential and falls through to the `priorValues` map. However, when `a` has no prior value and the optional subst returns nil, the concat still picks up the partially-resolved value from an earlier pass, yielding a doubled string.

## Test

`internal/resolver/resolver_test.go` — `TestSpecS13a_13_OptionalSelfRefUndefinedBecomesEmpty` (added in Phase 3 PR, currently skipped with `t.Skipf("spec violation, see #<this issue>")`).

## Labels

spec-violation

### PR DESCRIPTION
**Pull request by y1o1**
## Summary

Fix S13a.13: `a = ${?a}foo` with no prior `a` now resolves to `"foo"` (per HOCON.md L841), was `"foofoo"`.

Closes #68.

## Mechanism

When the resolver encounters a self-referential substitution `${?a}` (or `${a}`) inside the value being assigned to path `a`:
1. If a prior assignment to `a` exists → look-back returns the prior value (unchanged behavior).
2. **NEW**: If no prior exists → short-circuit. Optional `${?a}` → `nil` (concat-fold omits it); required `${a}` → `ResolveError`.

Identification: a substitution is "self-referential in the value being constructed" when **the looked-up value contains the exact same `substPlaceholder` AST node** as the substitution being resolved (`sp == s` pointer-identity check, vs. the prior path-equality check `slices.Equal(segTexts(...), segStrs)` which was too coarse).

Two-pronged round-2 fix:
1. **Fast-path cache check** at top of `resolveSubst`: if the path is fully resolved and cached, return immediately. Prevents double-expansion in forward-order (`a` first) cases.
2. **Pointer-identity isSelfRef**: replaces structural path-equality. Handles reverse-order (`b = ${a}; a = ${?a}foo`) cases via the existing `r.resolving["a"]==true` guard.

Bonus fix: `setPath` now populates `parent.priorValues[lastKey]` for nested-path single-segment assignments (single-line addition mirroring the existing line 282 pattern). Without this, sr10 (`foo.a = "x"; foo.a = ${?foo.a}bar`) returned `"bar"` instead of `"xbar"`.

## Why not the spec's path-equality?

The original spec ★1 decision #1 specified path-equality preservation. Round-2 multi-agent-review surfaced a false-positive: external lookups like `a = ${?a}foo; b = ${a}` would error because `b`'s `${a}` lookup found `a`'s concat-with-self-ref and triggered the short-circuit (wrongly).

All 3 impls (ts/rs/go) independently tightened the mechanism. ts uses node-membership-in-iterating-set; rs uses a path stack + `is_owner` guard; go uses this AST-node pointer-identity. **All three are strictly narrower than path-equality** (correctly exclude external lookups).

Spec amendment deferred to a follow-up xx.hocon PR. See [xx.hocon#27](https://github.com/o3co/xx.hocon/issues/27) for the cross-impl tracking issue.

> Note on commit subject: commit 14a4ddf's subject reads "tighten self-ref detection to current-field-path equality" — the body correctly describes pointer-identity. The subject was inherited from the original plan wording; the actual mechanism is pointer-identity (matching this PR description).

## Test plan

- [x] Conformance fixtures sr01–sr11 (xx.hocon `testdata/hocon/self-ref-lookback/`) all pass via `s13a13_self_ref_lookback_test.go`
- [x] Resolver-layer unit tests for 10+ edge cases (no-prior, with-prior, required, array, nested, mutual-ref)
- [x] Round-2 regression test `TestSpecS13a_13_ExternalReferenceToSelfRefField`: `a = ${?a}foo; b = ${a}` → both `"foo"` (forward + reverse declaration order)
- [x] Existing `TestSpecS13a_13_OptionalSelfRefUndefinedBecomesEmpty` un-skipped
- [x] Orphaned `specIssueS13a13` const removed
- [x] s13c_env_var_list_test.go tripwire comments updated
- [x] `go test ./...` green (~1.6k tests, `-count=1`)
- [x] `golangci-lint run ./...` clean (0 issues)
- [x] `gofmt -l .` clean
- [x] No exported API changes (`internal/resolver/` package)

## Pre-existing bugs surfaced (out of scope, see xx.hocon#27)

Round 2 review discovered 4 pre-existing resolver bugs beyond S13a.13 strict:
- Nested external ref crash (`foo.a = ${?foo.a}bar; foo.b = ${foo.a}`)
- Cache pollution on prior-with-external-ref
- Same-field double-self-ref crash
- Order-dependent ref-before-self-ref

These are filed as cross-impl follow-ups; not addressed in this PR.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: s13a13_self_ref_lookback_test.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+func expectedPath_s13a13(name string) string {
+	return filepath.Join("testdata", "expected", "self-ref-lookback", name+"-expected.json")
+}
+
+// TestS13a13_SelfRefLookback_Success asserts that each success fixture parses
+// and resolves to the expected JSON value.
+func TestS13a13_SelfRefLookback_Success(t *testing.T) {
+	if _, err := os.Stat(filepath.Join("testdata", "expected", "self-ref-lookback")); err != nil {
+		t.Skip("self-ref-lookback expected fixtures missing; run `make testdata`")
```
[the remark this answers is not in the document]
Comment (y1o1) [reply s13a13_self_ref_lookback_test.go:61]: Dismissed (verified-empirical): CI workflow `.github/workflows/test.yml` runs `make testdata` before `go test ./...`, and the Makefile's `cp -R "$tmpdir/expected/hocon/." "$(EXPECTED_DIR)/"` recursively syncs self-ref-lookback expected JSON. All previous Phase 6 clusters (include-reservation, env-var-list, units-default, concat-errors) follow this exact pattern with CI-green confirmation. This PR's CI is green on go1.23/24/25 across ubuntu/macos/windows, confirming tests actually exercise (not skip). Overturn condition: if CI logs ever show S13a.13 conformance tests SKIPPED (not RUN), reopen.

File: s13a13_self_ref_lookback_test.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	}
+}
+
+// TestS13a13_SelfRefLookback_Errors asserts that each error fixture produces
+// a non-nil error during parse or resolve.
+func TestS13a13_SelfRefLookback_Errors(t *testing.T) {
+	if _, err := os.Stat(filepath.Join("testdata", "expected", "self-ref-lookback")); err != nil {
+		t.Skip("self-ref-lookback expected fixtures missing; run `make testdata`")
+	}
+
```
[the remark this answers is not in the document]
Comment (y1o1) [reply s13a13_self_ref_lookback_test.go]: Fixed in 00e787f — separated the skip gate. `TestS13a13_SelfRefLookback_Errors` now checks the `.conf` file existence directly (which is committed in this PR), not the expected JSON directory. The error test runs unconditionally when its fixtures are present.

File: internal/resolver/resolver.go [hunk 3]
```diff
@@ -920,8 +953,11 @@ func (r *resolver) setPath(obj *ObjectVal, segments []string, val Val) {
 					val = deepMerge(nv, eo) // new over existing: nv=dst wins
 				}
 			} else {
-				// non-object overwrite: save prior value for self-referential substitution support
+				// non-object overwrite: save prior value for self-referential substitution support.
+				// Store on the parent object's priorValues so that nested-path self-ref
+				// look-back (resolveSubst isSelfRef → parent.priorValues check) can find it.
 				r.priorValues[segmentsToKey(segments)] = existing
+				obj.priorValues[key] = existing
```
[the remark this answers is not in the document]
Comment (y1o1) [reply internal/resolver/resolver.go]: Fixed in 00e787f — chose option (b): removed the `r.priorValues[segmentsToKey(segments)] = existing` write entirely from the nested `setPath` base case. The `parent.priorValues[key]` write retained for nested self-ref look-back is sufficient (resolveSubst reads both, but for nested cases only the parent entry is correct). Top-level keys are correctly handled by buildObject's `len==1` branch. Added regression test `TestS13a13_NestedPriorDoesNotCollideWithTopLevel` verifying `foo.a = ${?foo.a}bar; a = ${?a}zzz` doesn't cross-pollinate.
